### PR TITLE
feat: US135527 Add FAD Dates To Activity Usage Entity

### PR DESCRIPTION
**Pull request by pureooze**
This PR adds the FAD dates to the assignments entity in `siren-sdk`.
There will be a corresponding PR in `activities` to make use of the new properties to show the FAD date type picker.

## Rally Story
https://rally1.rallydev.com/#/?detail=/userstory/623042532951&fdp=true

### CODE REVIEW

File: src/hypermedia-constants.js [hunk 1]
```diff
@@ -347,6 +347,13 @@ export const Classes = {
 		issueDate: 'issue-date',
 		startDate: 'start-date'
 	},
+	availabilityDates: {
+		availabilityDates: 'availability-dates',
+		dueDate: 'due-date',
+		endDate: 'end-date',
+		issueDate: 'issue-date',
+		startDate: 'availability-start-date'
+	},
```
Comment (adkulas) [inline src/hypermedia-constants.js:356]: I don't think this maps over to the entity correctly, however, my knowledge around siren is limited.
Based on the [serializer](https://search.d2l.dev/xref/lms/le/activities/D2L.LE.Activities.Extensibility/Api/Hypermedia/ActivitiesSirenSerializers.cs?r=5bec958a#163-175) I think we only need:
- availabilityDates
- startDate
- endDate
```suggestion
	availabilityDates: {
		availabilityDates: 'availability-dates',
		endDate: 'availability-end-date',
		startDate: 'availability-start-date'
	},
```

File: src/activities/ActivityUsageEntity.js [hunk 2]
```diff
@@ -200,12 +200,43 @@ export class ActivityUsageEntity extends Entity {
...
+
+		return availabilityStartDate?.properties?.dateType;
+	}
+
+	/**
+	 * @returns {Number} Default start date type of the activity usage
+	 */
+	defaultStartDateType() {
+		const dateEntity = this._getSubEntityByClass(Classes.availabilityDates.availabilityDates);
+		return dateEntity?.getActionByName(Classes.availabilityDates.create)?.fields[1]?.value;
```
Comment (adkulas) [inline src/activities/ActivityUsageEntity.js:239]: Same as above, optional chaining may be a problem

File: src/activities/ActivityUsageEntity.js [hunk 3]
```diff
@@ -227,10 +258,33 @@ export class ActivityUsageEntity extends Entity {
...
+
+		return availabilityEndDate?.properties?.dateType;
+	}
+
+	/**
+	 * @returns {Number} Default start date type of the activity usage
+	 */
+	defaultEndDateType() {
+		const dateEntity = this._getSubEntityByClass(Classes.availabilityDates.availabilityDates);
+		return dateEntity?.getActionByName(Classes.availabilityDates.create)?.fields[3]?.value;
```
Comment (adkulas) [inline src/activities/ActivityUsageEntity.js:285]: Same as above, optional chaining may be a problem

File: src/activities/ActivityUsageEntity.js [hunk 4]
```diff
@@ -200,12 +200,43 @@ export class ActivityUsageEntity extends Entity {
...
 	/**
 	 * @returns {string} Start date of the activity usage
 	 */
 	startDate() {
-		const startDate = this._getDateSubEntity(Classes.dates.startDate);
-		return startDate && startDate.properties.date;
+		const availabilityStartDate = this._getAvailabilityStartDateEntity();
+
+		if (availabilityStartDate) {
+			return availabilityStartDate?.properties?.dateTime.date;
```
[the remark this answers is not in the document]
Comment (pureooze) [reply src/activities/ActivityUsageEntity.js]: Yeah I had to remove them all 😢

File: src/hypermedia-constants.js (hunk 1 above)
Comment (pureooze) [reply src/hypermedia-constants.js:356]: Ah yes, I copied the old dates one at the start because I didnt know what the schema would look like.

File: src/activities/ActivityUsageEntity.js [hunk 5]
```diff
@@ -181,7 +181,7 @@ export class ActivityUsageEntity extends Entity {
 	 * @returns {string} Due date of the activity usage
 	 */
 	dueDate() {
-		const dueDate = this._getDateSubEntity(Classes.dates.dueDate);
+		const dueDate = this._getSubEntityByClass(Classes.dates.dueDate);
```
[the remark this answers is not in the document]
Comment (pureooze) [reply src/activities/ActivityUsageEntity.js:184]: Yeah, I figured might as well just pass in the class you want 🤷‍♀️

File: src/activities/ActivityUsageEntity.js [hunk 6]
```diff
@@ -200,12 +200,62 @@ export class ActivityUsageEntity extends Entity {
...
+	 */
+	defaultStartDateType() {
+		const dateEntity = this._getSubEntityByClass(Classes.availabilityDates.availabilityDates);
+
+		if (!dateEntity) {
+			return;
+		}
+
+		const action = dateEntity.getActionByName(Classes.availabilityDates.create);
+		if (!action || !action.fields[1]) {
```
Comment (pureooze) [inline src/activities/ActivityUsageEntity.js:254]: Is there a better way to check the specific field on the action? I really really don't like doing it this way....

Comment (adkulas) [reply src/activities/ActivityUsageEntity.js:254]: There seems to be a `getFieldByName` method here https://search.d2l.dev/xref/BrightspaceHypermediaComponents/siren-sdk/src/activities/quizzes/timing/QuizTimingEntity.js?r=9bccd289#172

File: src/activities/ActivityUsageEntity.js [hunk 7]
```diff
@@ -200,12 +200,62 @@ export class ActivityUsageEntity extends Entity {
 		await this._setDate(dateValue, Classes.dates.dueDate, 'dueDate');
 	}
 
+	/**
+	 * @returns {boolean} If F18167-flexible-assignment-dates-ux feature is enabled
```
Comment (kieranderson) [inline src/activities/ActivityUsageEntity.js:204]: This is unrelated to the assignments, this is for all activities

Comment (pureooze) [reply src/activities/ActivityUsageEntity.js:204]: Good point, will update that. I was just testing using assignments.

File: src/activities/ActivityUsageEntity.js [hunk 8]
```diff
@@ -227,10 +277,47 @@ export class ActivityUsageEntity extends Entity {
 	 * @returns {string} End date of the activity usage
 	 */
 	endDate() {
-		const endDate = this._getDateSubEntity(Classes.dates.endDate);
+		const availabilityEndDate = this._getAvailabilityEndDateEntity();
+
+		if (availabilityEndDate) {
+			return availabilityEndDate;
```
Comment (kieranderson) [inline src/activities/ActivityUsageEntity.js:283]: Should this be returning `return availabilityEndDate.properties.dateTime.date;`?

**Review by kieranderson — CHANGES_REQUESTED** [review]
The EndDate return should be changed I think. Other than that looks good.

File: src/activities/ActivityUsageEntity.js (hunk 6 above)
Comment (pureooze) [reply src/activities/ActivityUsageEntity.js]: Updated it to use `getFieldByName`

File: src/activities/ActivityUsageEntity.js [hunk 9]
```diff
@@ -200,12 +200,61 @@ export class ActivityUsageEntity extends Entity {
...
+		}
+
+		const startDate = this._getSubEntityByClass(Classes.dates.startDate);
+
+		if (!startDate || !startDate.properties) {
+			return;
+		}
+
+		return startDate.properties.date;
+	}
```
Comment (adkulas) [inline src/activities/ActivityUsageEntity.js:228]: So I think there may some potential errors that come up from this

The way the entity is added in the backend is if you have date types on the usage, it will add the "activity-availability-dates" entity. However, that entity may not have the "availability-start-date" or "availability-end-date" sub entities if there are no dates set for those.

- [dates entity added here](https://search.d2l.dev/xref/lms/le/activities/D2L.LE.Activities.Extensibility/Api/Hypermedia/Serializers/ActivityUsageDecorator.cs?r=1d2e61c7#141)
- [conditionally adding the date entities here](https://search.d2l.dev/xref/lms/le/activities/D2L.LE.Activities.Extensibility/Api/Hypermedia/ActivitiesSirenSerializers.cs?r=5bec958a&fi=SerializeActivityAvailabilityDates#156-179)

TLDR;
you'll always have a dates entity but you're not guaranteed the date entity

```suggestion
		const availabilityDatesEntity = this._getSubEntityByClass(Classes.availabilityDates.availabilityDates);
		const availabilityStartDate = this._getAvailabilityStartDateEntity();

		if (availabilityDatesEntity ) {
			return availabilityStartDate && availabilityStartDate.properties.dateTime && availabilityStartDate.properties.dateTime.date;
		}

		const startDate = this._getSubEntityByClass(Classes.dates.startDate);

		if (!startDate || !startDate.properties) {
			return;
		}

		return startDate.properties.date;
	}
```

Comment (adkulas) [reply src/activities/ActivityUsageEntity.js:228]: The error I'm thinking about could be: 
if you have fad enabled, but you don't have a startDate it won't short circuit and would return the date from the startDate entity

File: src/activities/ActivityUsageEntity.js [hunk 10]
```diff
@@ -335,14 +423,39 @@ export class ActivityUsageEntity extends Entity {
...
+	_getAvailabilityStartDateEntity() {
+		const availabilityDatesEntity = this._getSubEntityByClass(Classes.availabilityDates.availabilityDates);
+
+		if (!availabilityDatesEntity) {
+			return;
+		}
+
+		return availabilityDatesEntity.getSubEntityByClass(Classes.availabilityDates.startDate);
+	}
+
```
Comment (adkulas) [inline src/activities/ActivityUsageEntity.js:435]: This may need to change as well since having the dates entity doesn't guarantee taht you'll have the date startDate sub entity

**Review by adkulas — CHANGES_REQUESTED** [review]

**Comment by adkulas** [comment]
@pureooze @kieranderson another thing I just thought about. In the backend work I added a property for `currentAvailabilityType` [source](https://search.d2l.dev/xref/lms/le/activities/D2L.LE.Activities.Extensibility/Api/Hypermedia/ActivitiesSirenSerializers.cs?r=5bec958a&fi=SerializeActivityAvailabilityDates#152). Is there something we would need to add to siren here to expose those on the face side? I figured it could turn out to be useful eventually

File: src/activities/ActivityUsageEntity.js (hunk 10 above)
Comment (adkulas) [reply src/activities/ActivityUsageEntity.js:435]: I guess this one is good depending on the expected behaviour:
- fad off would imply no dates entity which returns "undefined"
- fad on with no start date would also return "undefined"

File: src/activities/ActivityUsageEntity.js (hunk 9 above)
Comment (pureooze) [reply src/activities/ActivityUsageEntity.js:228]: Yeah good point, this would fall through in those cases. Will update.
